File: ios/AirChatKit/Sources/AirChatProtocol/Logging.swift
import Foundation
import os

/// Logging seam so the protocol core stays free of UI and platform frameworks.
public protocol AirChatLogger: AnyObject {
    func log(_ tag: String, _ message: String)
}

public final class NoopLogger: AirChatLogger {
    public init() {}
    public func log(_ tag: String, _ message: String) {}
}

/// Single shared os.Logger instance; `os.Logger` is cheap and privacy-aware.
public final class OsLogger: AirChatLogger {
    private let logger = Logger(subsystem: "app.airchat", category: "protocol")

    public init() {}

    public func log(_ tag: String, _ message: String) {
        logger.debug("[\(tag, privacy: .public)] \(message, privacy: .public)")
    }
}

/// Writes protocol logs to stderr, which is what `xcrun devicectl device process launch
/// --console` captures.
///
/// This is the only way to read a real iPhone's protocol logs from the host: the unified log is
/// not reachable over `devicectl`, and the in-app buffer requires tapping through the UI. Wired in
/// for debug builds only (see `AppContainer`).
public final class ConsoleLogger: AirChatLogger {
    public init() {}

    public func log(_ tag: String, _ message: String) {
        FileHandle.standardError.write(Data("AIRCHAT_LOG [\(tag)] \(message)\n".utf8))
    }
}

/// Bounded in-memory log tail, surfaced by the settings screen for on-device debugging.
public final class BufferLogger: AirChatLogger {
    private let capacity: Int
    private var entries: [String] = []

    public init(capacity: Int = 500) {
        self.capacity = capacity
    }

    public func log(_ tag: String, _ message: String) {
        entries.append("\(tag): \(message)")
        if entries.count > capacity {
            entries.removeFirst(entries.count - capacity)
        }
    }

    public func snapshot() -> [String] { entries }

    public func clear() { entries.removeAll() }
}

public final class FanOutLogger: AirChatLogger {
    private let sinks: [AirChatLogger]

    public init(_ sinks: AirChatLogger...) {
        self.sinks = sinks
    }

    public func log(_ tag: String, _ message: String) {
        for sink in sinks { sink.log(tag, message) }
    }
}

File: android/core-protocol/src/main/kotlin/com/airchat/protocol/Logging.kt
package com.airchat.protocol

/**
 * Minimal logging seam so the protocol core stays free of Android and Kotlin-logging deps.
 * Platform modules (core-ble / app) install a real implementation; tests use [NOOP].
 */
interface AirChatLogger {
    fun log(tag: String, message: String)

    companion object {
        val NOOP: AirChatLogger = object : AirChatLogger {
            override fun log(tag: String, message: String) = Unit
        }
    }
}

/** In-memory logger used by tests and the debug UI. */
class BufferLogger(private val capacity: Int = 500) : AirChatLogger {
    private val entries = ArrayDeque<String>()

    override fun log(tag: String, message: String) {
        entries.addLast("$tag: $message")
        while (entries.size > capacity) entries.removeFirst()
    }

    fun snapshot(): List<String> = entries.toList()

    fun clear() = entries.clear()
}

/** A logger that forwards to a platform sink while also keeping a bounded in-memory tail. */
class FanOutLogger(private vararg val sinks: AirChatLogger) : AirChatLogger {
    override fun log(tag: String, message: String) {
        for (sink in sinks) sink.log(tag, message)
    }
}
